### ISSUE TITLE
a crash

### ISSUE THREAD

**Issue by yangzhihao** [mutualmobile/MMWormhole#24]
hi,i find in sometimes the listenerBlock will call  a lot of times and then crash,is it i call pass message with same identifier too quickly so that the first is not save succeed?

**Comment by cnstoll** [comment]
Could be. Where is the crash happening?

If you flood WatchKit with UI update commands, especially when the watch interface controller is off screen, it's fairly likely that you'll cause the iPhone process "watchd" or whatever it's called that is responsible for updating the watch UI to crash. That is fairly reproducible especially while reloading tables.

You probably want to limit the frequency you're passing data back and forth, especially if UI updates are involved.

**Comment by frahar** [comment]
Hi,
I had similar issues. For me it happened as I was calling "listenForMessageWithIdentifier" for the same identifier several times during the app runtime.
Can it be "registerForNotificationsWithIdentifier" shouldn't be called more than once for a single identifier?


For me modifying the MMWormHole method in the following way removed the multiple triggers.
As I am not used to GIT, I post this here and not directly into the code base
```
- (void)listenForMessageWithIdentifier:(NSString *)identifier
                              listener:(void (^)(id messageObject))listener {
    if (identifier != nil) {
        BOOL wasRegisteredBefore = (self.listenerBlocks[identifier] != nil);
        [self.listenerBlocks setValue:listener forKey:identifier];
        if (!wasRegisteredBefore) {
            [self registerForNotificationsWithIdentifier:identifier];
        }
    }
}
```

**Comment by HiddenJester** [comment]
I had a similar problem in a Today extension: if the extension had been run, then Today was closed, then my app posted a wormhole message, then Today was opened again (before the extension had unloaded), I'd get get a crash about an Optional unexpectedly being nil (this is in Swift code) in the block provided to listenForMessageWithIdentifier. Instead of tweaking the wormhole code I just created a bool to track if I had registered the listener already.

Maybe that's confusing. What I'm saying is if the today widget registered a listener, then the view was destroyed and the widget was suspended, then it reactivated and attempted to register the listener it would crash. I'm not sure if it processed the message and THEN registered the listener or registered the listener first then crashed during the processing. 

Here's my working code (with the bool flag):
```
	func registerListeners() {
		if registered {
			return
		}

		wormhole.listenForMessageWithIdentifier(appUpdatedKey) { (message) -> Void in
			processUpdatedWormholeMessage(message)
		}

		// If a message already exists in the wormhole go ahead and pick it up now.
		if let pinnedMessage: AnyObject = wormhole.messageWithIdentifier(appUpdatedKey) {
			processUpdatedWormholeMessage(pinnedMessage)
		}

		wormhole.listenForMessageWithIdentifier(appDeletedKey) { (message) -> Void in
			processDeleteWormholeMessage(message)
		}

		// If a message already exists in the wormhole go ahead and pick it up now.
		if let pinnedMessage: AnyObject = wormhole.messageWithIdentifier(appDeletedKey) {
			processDeleteWormholeMessage(pinnedMessage)
		}
		registered = true
	}
```

without the check for if registered I would crash in the first listen block - on processUpdatedWormholeMessage(message)

**Comment by HiddenJester** [comment]
I should mention that I'm no longer sure my problem was because of multiple calls to listenForMessageWithIdentifier. I've found that it is possible to receive a message this is null and this will cause the crash about the Optional unexpectedly being nil. I just submitted a [pull request](https://github.com/mutualmobile/MMWormhole/pull/34) that uses the nullability annotations to force Swift to see that message in the listener block isn't AnyObject!, but in fact AnyObject?

The pull request has more information but I strongly suggest anyone using MMWormhole with Swift write their listener blocks like this:

````
wormhole.listenForMessageWithIdentifier(appToWidgetUpdatedKey) { (message) -> Void in
        if let message: AnyObject = message {
            processUpdatedWormholeMessage(message)
        }
    }
`````

The pull request doesn't remove the need to write that, it just means Swift will insist on some sort of optional unwrap on message in the block handler.
